### PR TITLE
schema: first pass at parsing from JSON

### PR DESCRIPTION
**Pull request by emberian**
Note that I converted some paths to be Strings already,
in preparation for using JSON Pointers instead of absolute
JSON paths.

### CODE REVIEW

File: src/schema.rs [hunk 1]
```diff
@@ -4,18 +4,74 @@
...
+use self::error::ErrorKind::InvalidSchema;
 
 /// Schema
 pub struct Schema {
     files: BTreeMap<String, File>,
 }
 
+fn filename_is_valid(s: &str) -> bool {
+    lazy_static !{
+        static ref RE: Regex = Regex::new("([[:alnum:]]|_)+").expect("failed compiling regex!");
```
Comment (jviotti) [inline src/schema.rs:32]: What about names with hyphens?

File: src/schema.rs [hunk 2]
```diff
@@ -28,19 +84,117 @@ pub struct File {
...
+    expect_object(get(v, k)?)
+}
+
+fn get_string<'a>(v: &'a Value, k: &str) -> error::Result<&'a str> {
+    match get(v, k)?.as_str() {
+        Some(s) => Ok(s),
+        None => bail!(InvalidSchema(format!("expected a string for {}", k))),
+    }
+}
+impl File {
```
Comment (jviotti) [inline src/schema.rs:139]: nit: missing a new line above here :)

File: src/schema.rs [hunk 3]
```diff
@@ -55,21 +209,124 @@ pub enum Location {
...
 }
 
+impl Location {
+    pub fn from_json(v: &Value) -> error::Result<Location> {
+        let o = expect_object(v)?;
+        match o.get("parent") {
+            Some(p) => {
+                Ok(
+                    Location::Dependent {
+                        location: expect_string(p)?.to_owned(),
```
Comment (jviotti) [inline src/schema.rs:219]: Keep in mind that a dependent location also contains a property: https://github.com/resin-io/reconfix/blob/rust/docs/schema.markdown#string-property

**Review by jviotti — COMMENTED** [review]
I left some comments, but looks good overall. Should we have some unit tests though?

File: src/schema.rs [hunk 4]
```diff
@@ -4,18 +4,74 @@
...
+}
+
 /// Supported output file formats
 pub enum FileFormat {
     Ini,
     Json,
 }
 
+impl FileFormat {
+    pub fn from_str(s: &str) -> error::Result<FileFormat> {
```
Comment (abrodersen) [inline src/schema.rs:67]: Couldn't this be an impl of the `FromStr` trait?

**Review by abrodersen — COMMENTED** [review]
A good start, tests would be useful.

File: src/schema.rs (hunk 1 above)
Comment (emberian) [reply src/schema.rs:32]: According to the schema spec, that is not valid ;)

> Must consist of alphanumeric characters and underscores only

File: src/schema.rs (hunk 4 above)
Comment (emberian) [reply src/schema.rs:67]: Indeed!

File: src/schema.rs (hunk 3 above)
Comment (emberian) [reply src/schema.rs:219]: Shouldn't that be `String[] path`?

**Comment by jviotti** [comment]
What are we missing from this PR, so that we can fully unblock https://github.com/resin-io/reconfix/pull/30?

**Comment by jviotti** [comment]
Let's focus on finishing this PR so we can get unblocked. @cmr Do you think you'll have time to finish it soon? Otherwise maybe @abrodersen can take over?

**Pull request merged**